vncviewer/Win32TouchHandler.cxx

```diff
@@ -0,0 +1,441 @@
+}
+
+void Win32TouchHandler::fakeMotionEvent(const GestureEvent origEvent)
+{
+  UINT Msg = WM_MOUSEMOVE;
+  WPARAM wParam = MAKEWPARAM(fakeButtonMask, 0);
+  LPARAM lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+
+  pushFakeEvent(Msg, wParam, lParam);
+  lastFakeMotionPos.x = origEvent.event_x;
+  lastFakeMotionPos.y = origEvent.event_y;
+}
+
+void Win32TouchHandler::fakeButtonEvent(bool press, int button,
+                                        const GestureEvent origEvent)
+{
+  UINT Msg = 0;
+  WPARAM wParam = MAKEWPARAM(0, 0);
+  LPARAM lParam = MAKELPARAM(0, 0);
+  int delta = 0;
+
+  switch (button) {
+
+  case 1: // left mousebutton
+    if (press) {
+      Msg = WM_LBUTTONDOWN;
+      fakeButtonMask |= MK_LBUTTON;
+    } else {
+      Msg = WM_LBUTTONUP;
+      fakeButtonMask &= ~MK_LBUTTON;
+    }
+    break;
+  case 2: // middle mousebutton
+    if (press) {
+      Msg = WM_MBUTTONDOWN;
+      fakeButtonMask |= MK_MBUTTON;
+    } else {
+      Msg = WM_MBUTTONUP;
+      fakeButtonMask &= ~MK_MBUTTON;
+    }
+    break;
+  case 3: // right mousebutton
+    if (press) {
+      Msg = WM_RBUTTONDOWN;
+      fakeButtonMask |= MK_RBUTTON;
+    } else {
+      Msg = WM_RBUTTONUP;
+      fakeButtonMask &= ~MK_RBUTTON;
+    }
+    break;
+
+  case 4: // scroll up
+    Msg = WM_MOUSEWHEEL;
+    delta = WHEEL_DELTA;
+    break;
+  case 5: // scroll down
+    Msg = WM_MOUSEWHEEL;
+    delta = -WHEEL_DELTA;
+    break;
+  case 6: // scroll left
+    Msg = WM_MOUSEHWHEEL;
+    delta = -WHEEL_DELTA;
+    break;
+  case 7: // scroll right
+    Msg = WM_MOUSEHWHEEL;
+    delta = WHEEL_DELTA;
+    break;
+
+  default:
+    vlog.error(_("Invalid mouse button %d, must be a number between 1 and 7."),
+               button);
+    return;
+  }
+
+  if (1 <= button && button <= 3) {
+    wParam = MAKEWPARAM(fakeButtonMask, 0);
+
+    // Regular mouse events expect client coordinates
+    lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+
+  } else if (4 <= button && button <= 7) {
+    POINT pos;
+
+    // Only act on wheel press, not on release
+    if (!press)
+      return;
+
+    wParam = MAKEWPARAM(fakeButtonMask, delta);
+
+    // Wheel events require screen coordinates
+    pos.x = (LONG)origEvent.event_x;
+    pos.y = (LONG)origEvent.event_y;
+
+    ClientToScreen(hWnd, &pos);
+    lParam = MAKELPARAM(pos.x, pos.y);
+  }
+
+  pushFakeEvent(Msg, wParam, lParam);
+}
+
+void Win32TouchHandler::fakeKeyEvent(bool press, int keysym,
+                                     const GestureEvent origEvent)
+{
+  UINT Msg = press ? WM_KEYDOWN : WM_KEYUP;
+  WPARAM wParam;
+  LPARAM lParam;
+  int vKey;
+  int repeatCount = 1;
+  int scanCode;
+  int extendedKey = 0;
+  int contextCode = 0;
+  int previousKeyState;
+  int transitionState = press ? 0 : 1;
+
+  switch(keysym) {
+
+  case XK_Control_L:
+    vKey = VK_CONTROL;
+    scanCode = 0x1d;
+    if (press) {
+      previousKeyState = fakeButtonMask & MK_CONTROL;
```

Comment:
This looks very weird and pointless. What is it supposed to do?

Reply:
[Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-keydown) say:

> The previous key state. The value is 1 if the key is down before the message is sent, or it is zero if the key is up.

This was an attempt at following that. Anyway, reverted to always setting previousKeyState to 1 on keydown.